[Sep-24] Add tests for `pending_`, `completed` and `stellar_transaction_id`

CassioMG opened the pull request:
This PR adds the below tests for Sep-24 `pending` and `completed` transactions as well as fetching transactions through the `stellar_transaction_id` query parameter:
- `has proper 'pending_' deposit transaction schema on /transaction`
- `has proper 'completed' deposit transaction schema on /transaction`
- `has proper 'pending_user_transfer_start' withdraw transaction schema on /transaction`
- `has proper 'completed' withdraw transaction schema on /transaction`
- `returns valid deposit transaction when using 'stellar_transaction_id' param`
- `returns valid withdraw transaction when using 'stellar_transaction_id' param`

![Screenshot 2023-07-04 at 20 37 32](https://github.com/stellar/stellar-anchor-tests/assets/3228151/82197969-4599-497b-b000-fdc8a57b6a02)

In order to fulfill those tests `Anchors will need to upload a config file` in the format below with credentials and info related to existing pending and completed transactions. In case the config file is not provided, the new tests will fail but the existing (old) tests should still pass which means it's a 'backwards compatible' change.
```
{
  "account": {
    "publicKey": "GDTW75HOEQFPAMZ7YTFRN2TMVLVBXK7QYC4AJMXISV2WLD7E7RSZTDIA",
    "secretKey": "SAMISXU66LL2QDOP3A4EEKIEV4BV7EG3BJYFR65OF6WBOM6IQTDQOKWG"
  },
  "depositPendingTransaction": {
    "id": "252062",
    "status": "any pending_ status"
  },
  "depositCompletedTransaction": {
    "id": "251876",
    "status": "completed",
    "stellar_transaction_id": "6d51bff72685b3d6a9a951ff695e7cb361865470011447a9f29f81cc2939c445"
  },
  "withdrawPendingUserTransferStartTransaction": {
    "id": "251986",
    "status": "pending_user_transfer_start",
    "amount_in": "222.00",
    "amount_in_asset": "stellar:ARST:GB7TAYRUZGE6TVT7NHP5SMIZRNQA6PLM423EYISAOAP3MKYIQMVYP2JO",
    "withdraw_anchor_account": "GDMSN2PQ3EB32LZY5JVACI4H7GUVQ3YUWOM4437IDTVQHHWHYG7CGA5Z",
    "withdraw_memo": "251986",
    "withdraw_memo_type": "text"
  },
  "withdrawCompletedTransaction": {
    "id": "251938",
    "status": "completed",
    "stellar_transaction_id": "7d9f932216d2476ecd6278c00c1366f2740a40ddbea0e4ec5b5615664592b4ce"
  }
}
```
![Screenshot 2023-07-04 at 20 39 30](https://github.com/stellar/stellar-anchor-tests/assets/3228151/7600fa2b-b9c4-4fee-9f09-e9e77c8a2d26)
![Screenshot 2023-07-31 at 18 37 46](https://github.com/stellar/stellar-anchor-tests/assets/3228151/b5a3a79a-620e-4380-b949-075284fd929a)

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

Ifropc commented on @stellar/anchor-tests/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@stellar/anchor-tests",
-  "version": "0.5.11",
+  "version": "0.5.12",
```
Can you set version to 0.6 please? 
AP uses 0.5.11 and if there's a hot fix necessary for current tests it'd be hard to fix them
(AP should eventually move to 0.6 but ideally we want to generate transactions from current build of the platform, so it could take a while)

Ifropc commented on @stellar/anchor-tests/src/schemas/sep24.ts:
```diff
@@ -1,3 +1,123 @@
+  },
+};
+
+const completedDepositConfigSchema = {
```
Can we make all this schemas more strict?
[Deposit](https://github.com/stellar/kotlin-wallet-sdk/blob/main/wallet-sdk/src/main/kotlin/org/stellar/walletsdk/anchor/AnchorTransaction.kt#L40)
[Withdrawal](https://github.com/stellar/kotlin-wallet-sdk/blob/main/wallet-sdk/src/main/kotlin/org/stellar/walletsdk/anchor/AnchorTransaction.kt#L66)
Both pending and completed. 
Everything that is not nullable should be mandatory.
This schema matches SEP-24 protocol and polaris implementation

CassioMG replied:
Sure, let me review it

CassioMG replied:
@Ifropc I think you were looking at the Config File schema, if you look at the Transaction schema you'll see it's already as strict as the kotlin one. I realized the schema file was looking a bit confusing so I just reorganized/rewrote some stuff to make it more clear. I've just put the Transaction schema at the top of the file and the Config File schema at the bottom of the file to avoid confusion. Please let me know if it's looking good now - thanks!

CassioMG replied:
I've left some extra comments to highlight them

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

CassioMG commented on @stellar/anchor-tests/src/schemas/sep24.ts:
```diff
@@ -99,90 +77,323 @@ export const transactionSchema = {
-    withdraw_anchor_account: {
-      type: ["string", "null"],
+// =========================== > CONFIG FILE schema < =========================== //
+const initialConfigSchema = {
```
Config file schema starts here

JakeUrban commented on @stellar/anchor-tests/src/cli.ts:
```diff
@@ -72,11 +72,12 @@ const command = yargs
       if (
         (argv.seps.includes(6) ||
           argv.seps.includes(12) ||
+          argv.seps.includes(24) ||
           argv.seps.includes(31) ||
           argv.seps.includes(38)) &&
         !argv.sepConfig
       ) {
-        throw "error: SEP 6, 12, 31 & 38 require a configuration file (--sep-config, -c)";
+        throw "error: SEP 6, 12, 24, 31 & 38 require a configuration file (--sep-config, -c)";
       }
```
Is it possible to make the config file optional here? In the UI it sounds like we make it optional and just fail the tests that require the config file, which I think is fine.

JakeUrban commented on @stellar/anchor-tests/src/helpers/config.ts:
```diff
@@ -164,6 +164,13 @@ function checkSepConfigObj(config: Config) {
       );
     }
   }
+  if (config.seps.includes(24)) {
+    if (!config.sepConfig || !config.sepConfig["24"]) {
+      throw new ConfigError(
+        "SEP 24 configuration is required to run SEP-24 tests.",
+      );
+    }
+  }
```
Why do we need this check if the config file is optional in the UI?

CassioMG replied:
We can probably remove it, I'll take look

JakeUrban commented on @stellar/anchor-tests/src/types.ts:
```diff
@@ -106,6 +107,89 @@ export interface Sep12Config {
+   *
+   * When provided, it'll be used in the sep10 "returns a valid JWT" test.
+   */
+  accountHolder: {
+    accountAddress: string;
+    accountSignerSecretKey: string;
+  };
```
SEP-31's config object has something similar, `sendingAnchorClientSecret`. We don't actually need the public key configured right? Could we do something like `authAccountSecret`?

CassioMG replied:
So in most cases we will only need the secret key as we can infer the public key from it. But in certain cases (like Vibrant accounts) we have a public key with weigh zero so we need a separate secret key to actually sign stuff. We can probably make the `accountAddress` optional and in case it's not provided then we infer the public key using the provided secret key. How does that sound?

About the naming, I thought about using `accountSigner...` to make it clear we are talking about a signer of the account in question. But, as in most cases Anchors will only need to pass the secretKey maybe we should make it less confusing for the common case.

What do you think about renaming the whole object like the below ?
```
account: {
    publicKey?: string;
    secretKey: string;
  };
```

JakeUrban replied:
I like that approach!

CassioMG replied:
Done!

JakeUrban commented on ui/src/components/ConfigModalContent/index.tsx:
```diff
@@ -151,6 +152,122 @@ export const ConfigModalContent: React.FC = () => (
       </ul>
       <p>Each customers key assigned to the other attributes must be unique.</p>
 
+      <Heading4>SEP-24</Heading4>
+      <p>
+        SEP-24 configuration objects are used for specifying the account
+        credentials and transaction parameters that should be used when making
+        GET /transaction requests for fetching pending and completed
+        transactions.
+      </p>
```
Can we note that the config file is optional if the anchor doesn't want to set up transactions in these statuses before running tests?

CassioMG replied:
Done

JakeUrban commented on @stellar/anchor-tests/src/helpers/sep24.ts:
```diff
@@ -0,0 +1,68 @@
+export const invalidConfigFile: Failure = {
+  name: "invalid config file",
+  text(args: any): string {
+    return "The UPLOAD CONFIG file has some issues:\n\n" + `${args.errors}.`;
```
Can we update this message to account for the possibility that the config file wasn't provided?

CassioMG replied:
I'll update the message

CassioMG replied:
Just pushed this change, this is how it'll look like when the Sep-24 config object is missing:
![Screenshot 2023-07-31 at 16 24 35](https://github.com/stellar/stellar-anchor-tests/assets/3228151/3c0b9b62-ac12-4d41-831a-c68709355acf)

JakeUrban reviewed (commented):
Thanks for the PR @CassioMG! I appreciate some the refactoring, and I'm curious what your thoughts are on the library as a whole compared.

My main request is that we make these tests optional in both the UI and CLI. These are nice tests to have but require the anchor to setup these transactions on their backend prior to running tests, and I don't want folks to be blocked by that if they just want to run the other tests initially.

Ideally the testrunner would skip the tests that require the config file if it wasn't specified, but thats a larger change (supporting optional tests) so we move forward with having them fail for now.

CassioMG commented:
@JakeUrban thanks for reviewing it! Appreciate the comments

> Thanks for the PR @CassioMG! I appreciate some the refactoring, and I'm curious what your thoughts are on the library as a whole.

I didn't extensively studied the library but as far as I can tell it looks good. It has some well defined modules and things are well organized in files and folders so it helps when you simply need to attach new tests or modules.

> My main request is that we make these tests optional in both the UI and CLI. These are nice tests to have but require the anchor to setup these transactions on their backend prior to running tests, and I don't want folks to be blocked by that if they just want to run the other tests initially.

This PR implements it in a way that - in reality - the config file is already optional. Like if the Anchor doesn't upload any config file an empty dummy config file will be used instead, which will result in all (old) existing tests to pass but the new ones to fail. Which means Anchors are able to run the other tests initially without providing the config file (they can just tap "run" and things will work).

I agree that it could be better communicated on the copies to make it clear to Anchors that it's optional, I'll work on that.

> Ideally the testrunner would skip the tests that require the config file if it wasn't specified, but thats a larger change (supporting optional tests) so we move forward with having them fail for now.

I actually thought about leaving the new config-required tests failing on the screen so it could encourage Anchors to take the necessary steps for the new tests to succeed. If Anchors manage to make the new tests succeed it would reduce considerably the burden of wallets having to educate Anchors on how to properly manage pending and completed transactions.

Probably a better approach would be displaying some yellow warning signs ⚠️ next to the config-required tests in case the config file is not provided so Anchors could pay attention to it but understand it's not _critically_ required instead of displaying the failed red ❌ signs. Does that sound good to you?

JakeUrban commented:
> Probably a better approach would be displaying some yellow warning signs ⚠️ next to the config-required tests in case the config file is not provided so Anchors could pay attention to it but understand it's not critically required instead of displaying the failed red ❌ signs. Does that sound good to you?

That sounds great 👍

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

JakeUrban reviewed (approved):
Thanks @CassioMG, LGTM

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

stellar-jenkins commented:
**Preview is available here:**<br>http://stellar-anchor-tests-pr117.previews.kube001.services.stellar-ops.com/

CassioMG commented:
> Probably a better approach would be displaying some yellow warning signs ⚠️ next to the config-required tests in case the config file is not provided so Anchors could pay attention to it but understand it's not _critically_ required instead of displaying the failed red ❌ signs. Does that sound good to you?

@JakeUrban hey Jake, I've addressed all comments from this PR except my suggestion of adding this new warning ⚠️  status. I realized now that this is a bigger change, it won't be trivial to add an intermediary "warn" status in between "success" and "fail". This would also require some work on [@stellar/design-system](https://github.com/stellar/stellar-design-system/tree/main/%40stellar/design-system) in order for the anchor-tests to display a consistent UI for this new yellow status. So my proposal is to merge it like this (e.g. display errors with the "missing Sep-24 config file" message on the new tests to give guidance to Anchors) and tackle the "warn" status later as a separate enhancement task. Does that sound good to you?

JakeUrban commented:
Yes, sounds good to me.
